### PR TITLE
Integration Cerberus 

### PR DESCRIPTION
**Pull request by vipulgupta2048**
This PR marks the supposed ending of Cerberus's long journey into integration with its other pals.

Signed-off-by: Vipul Gupta (@vipulgupta2048) <vipulgupta2048@gmail.com>

### CODE REVIEW

File: spidermon/contrib/scrapy/pipelines.py [hunk 1]
```diff
@@ -94,8 +105,17 @@ def _load_jsonschema_validator(cls, schema):
                 "- an object path to a JSON string.\n"
                 "- a path to a JSON file."
             )
+
         return JSONSchemaValidator(schema)
 
+    @classmethod
+    def _load_cerberus_validator(cls, schema):
+        if isinstance(schema, six.string_types):
```
Comment (ejulio) [inline spidermon/contrib/scrapy/pipelines.py:104]: Better to make these validations inside `__init__`.
Otherwise we need to carry them around whenever we want to use `CerberusValidator`

**Review by ejulio — CHANGES_REQUESTED** [review]

File: spidermon/contrib/scrapy/pipelines.py (hunk 1 above)
Comment (vipulgupta2048) [reply spidermon/contrib/scrapy/pipelines.py:104]: The `__init__` of `pipelines.py`??
As per my understanding of slack messages, I thought we were keeping both the validations. 
`NotConfigured` on the outside, and SchemaError in the side, for the off chance if we ever need to pick the `CerberusValidator()` we can do that without worrying about validating our schemas.

Comment (ejulio) [reply spidermon/contrib/scrapy/pipelines.py:104]: We need to check if the schema in `CerberusValidator`.
That's for sure, now, `CerberusValidator` shouldn't rase `NotConfigured` if it is invalid, maybe `SchemaError` is better.
So, here you can catch the given exception and raise `NotConfigured`.
The thing is that knowing if the given schema is valid is not something that `ItemValidationPipeline` should do, because the validation might vary among validators.

Comment (vipulgupta2048) [reply spidermon/contrib/scrapy/pipelines.py:104]: > The thing is that knowing if the given schema is valid is not something that ItemValidationPipeline should do, because the validation might vary among validators.

Yes @ejulio this makes sense. `ItemValidationPipelines` shouldn't have one endall method to check the schema for all validators, rather custom checking could be the best option here.

File: spidermon/contrib/scrapy/pipelines.py [hunk 2]
```diff
@@ -3,19 +3,28 @@
...
+    SchematicsValidator,
+    JSONSchemaValidator,
+    CerberusValidator,
+)
 from spidermon.contrib.validation.jsonschema.tools import get_schema_from
 from schematics.models import Model
 
 from .stats import ValidationStatsManager
-
+# import ipdb
```
Comment (ejulio) [inline spidermon/contrib/scrapy/pipelines.py]: Remove comments

File: spidermon/contrib/scrapy/pipelines.py [hunk 3]
```diff
@@ -31,7 +40,7 @@ def __init__(
         add_errors_to_items=DEFAULT_ADD_ERRORS_TO_ITEM,
         errors_field=None,
     ):
-        self.drop_items_with_errors = drop_items_with_errors
+        self.drop_items_with_errors = drop_items_with_errors or DEFAULT_DROP_ITEMS_WITH_ERRORS
```
Comment (ejulio) [inline spidermon/contrib/scrapy/pipelines.py]: Why this change?
The default value of `drop_items_with_errors` is already `DEFAULT_DROP_ITEMS_WITH_ERRORS`

File: spidermon/contrib/scrapy/pipelines.py (hunk 1 above)
Comment (ejulio) [reply spidermon/contrib/scrapy/pipelines.py:104]: Shouldn't these validations go to `CerberusValidator.__init__`?

File: tests/contrib/scrapy/test_pipeline_cerberus.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+STATS_TYPES = "spidermon/validation/validators/{}/{}"
+
+SETTING_CERBERUS = "SPIDERMON_VALIDATION_CERBERUS"
+
+def assert_type_in_stats(validator_type, obj):
+    return "'{}' in {{stats}}".format(
+        STATS_TYPES.format(obj.__name__.lower(), validator_type)
+    )
+
+ASSERT_TYPE_IN_STATS = partial(assert_type_in_stats, "cerberus")
```
Comment (ejulio) [inline tests/contrib/scrapy/test_pipeline_cerberus.py]: Does it really need to be a `partial`?
Wouldn't be easy to assume `cerberus` in `assert_type_in_stats`?

File: tests/contrib/scrapy/test_pipeline_cerberus.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            id="missing required fields",
+        ),
+    ],
+)
+def test_get_crawler_only(item, settings, cases):
+    crawler = get_crawler(settings_dict=settings)
+    pipe = ItemValidationPipeline.from_crawler(crawler)
+    pipe.process_item(item, None)
+    kwargs = {"stats":"pipe.stats.stats.get_stats()"}
+    for case in cases if type(cases) in [list, tuple] else [cases]:
```
Comment (ejulio) [inline tests/contrib/scrapy/test_pipeline_cerberus.py]: Easier if `cases` is always a `list` or `tuple` in the test case argument.
If not, at least put it in a variable, this `for` statement is quite long.

File: tests/contrib/scrapy/test_pipeline_cerberus.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        assert eval(case.format(**kwargs))
```
Comment (ejulio) [inline tests/contrib/scrapy/test_pipeline_cerberus.py]: `eval` is not good idea

Comment (ejulio) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: If you need to test an expression, then a lambda is completely fine.
Otherwise, you can create other test methods

**Review by ejulio — CHANGES_REQUESTED** [review]

File: spidermon/contrib/scrapy/pipelines.py (hunk 1 above)
Comment (vipulgupta2048) [reply spidermon/contrib/scrapy/pipelines.py:104]: Like we discussed, I guess. These validations are the same validations that are already present in `CerberusValidator` as the countermeasure to something that you said in regards to making code independent. 

As if someone would like to take our module then it should be independent and self-sustainable to use (Pardon my code vocabulary here :see_no_evil: )

Comment (vipulgupta2048) [reply spidermon/contrib/scrapy/pipelines.py:104]: Or maybe I got it wrong, here's the `__init__` method of `CerberusValidator` class

```
   def __init__(self, schema, translator=None, use_default_translator=True):
        super(CerberusValidator, self).__init__(
            translator=translator, use_default_translator=use_default_translator
        )
        if isinstance(schema, Mapping):
            self._schema = schema
        else:
            raise SchemaError("Validation schema missing")
```

Comment (vipulgupta2048) [reply spidermon/contrib/scrapy/pipelines.py:104]: Oooh, I see it now. I see you are referring to the `six.string_types` validation? 

Yes, that isn't there. And probably could be there. That's a good idea. Coming right up.

File: spidermon/contrib/scrapy/pipelines.py (hunk 2 above)
Comment (vipulgupta2048) [reply spidermon/contrib/scrapy/pipelines.py]: Yeah, just checked. Do apologize for missing this one.

File: spidermon/contrib/scrapy/pipelines.py (hunk 1 above)
Comment (ejulio) [reply spidermon/contrib/scrapy/pipelines.py:104]: Also, but if the validation is already in `CerberusValidator.__init__` why repeat it here?
Also, in `__init__` it is better to make the validations/assertions before any logic, in that case, calling `super`

File: spidermon/contrib/scrapy/pipelines.py (hunk 3 above)
Comment (vipulgupta2048) [reply spidermon/contrib/scrapy/pipelines.py]: I thought as `add_items_with_errors` is also structured like this then why not do this the same as well ... Quite honestly, didn't think about this change at that point. Will change both of them.

Comment (vipulgupta2048) [reply spidermon/contrib/scrapy/pipelines.py]: Changed both, resolving. Let me know otherwise if there is any other change. 
Resolving.

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 4 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: Well, this was part of my brainstorming questions that we needed to discuss but didn't ask.

`partial` has been used in `test_pipelines.py` to kinda delay the first argument's input just because in one class the `validator_type` is  `jsonschema` and the other its `schematics`.

My question here was:
I was earlier thinking of converting this to a fixture. What say about that? It seemed like overkill hence didn't do it. Later I thought about hardcoding which is clean but then we needed the obj so functions are a way to go.   
Since for this specific test. Cerberus will always be there hence any of the options seemed like an okay choice. Hence, I stuck with `partial` just because that's what the other tests did. 

Ps. Partial is pretty neat only when used in the right place. Didn't know about this.
Can transform this into a function? Let me know.

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 5 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: Picked it from the previous tests, found it quite complex to understand in the first go. 
Will make this simpler.

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 6 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: Discussion is needed on this bit. Will get it resolved in the meeting

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 4 above)
Comment (ejulio) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: For me, since we're only dealing with Cerberus, there is no need for an argument.
So, there is no need for `partial`;
Also, a `fixture` for this does not make sense, `fixtures` are for test inputs/state, not assertions.

Comment (ejulio) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: It is completely fine for a function to be `assert_type_in_stats` and set cerberus as constant here.
At least, also, when using `partial` is not a case to set the function `UPPER_CASE` like a constant, it is just a function. So, a better name would be `assert_in_cerberus_stats` for example

Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: Done and roger that! Good change, with the naming conventions!

Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: I keep following Pylint, now that I think of it. Maybe following too much of set of conventions can be counterproductive.

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 6 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: Lambda coming right up!

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 4 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: Done! :heavy_check_mark:

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 5 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: Finished this as well.

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 6 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: Opted for Lambda, looking much better now. Thanks for the suggestion!

Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: Added question about implementation on the Slack Channel.

File: spidermon/contrib/scrapy/pipelines.py [hunk 7]
```diff
@@ -56,6 +64,7 @@ def set_validators(loader, schema):
         for loader, name in [
             (cls._load_jsonschema_validator, "SPIDERMON_VALIDATION_SCHEMAS"),
             (cls._load_schematics_validator, "SPIDERMON_VALIDATION_MODELS"),
+            (cls._load_cerberus_validator, "SPIDERMON_VALIDATION_CERBERUS"),
```
Comment (ejulio) [inline spidermon/contrib/scrapy/pipelines.py:60]: I'm not so sure about the naming here.
Cerberus works with both models and schemas.
Maybe, we could set that for schemas the default is json schema.
For models, the default is schematics.
Then we could add a new setting `SPIDERMON_VALIDATION_FRAMEWORK` and the user can set it as the underlying framework. Cerberus would be an option.

@rennerocha , thoughts?

File: tests/contrib/scrapy/test_pipeline_cerberus.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    ],
+)
+def test_get_crawler_only(item, settings, cases):
+    crawler = get_crawler(settings_dict=settings)
+    pipe = ItemValidationPipeline.from_crawler(crawler)
+    pipe.process_item(item, None)
+    kwargs = {"stats": "pipe.stats.stats.get_stats()"}
+    for case in cases:
+        pass
+        # FIXX THISS
```
Comment (ejulio) [inline tests/contrib/scrapy/test_pipeline_cerberus.py]: If you want to add a comment, add a comment to the PR.
Do not push the code with these comments

File: tests/contrib/scrapy/test_pipeline_cerberus.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    settings = {SETTING_CERBERUS: [cerberus_error_test_schema]}
+    #     TestItem: {ScrapingHub"}
+    # }
+    test_item = {"url":"asd", "title":"wdad"}
+    crawler = get_crawler(settings_dict=settings)
+    pipe = ItemValidationPipeline.from_crawler(crawler)
+    pipe.process_item(test_item, None)
+    kwargs = {"stats": "pipe.stats.stats.get_stats()"}
+    stats = pipe.stats.stats.get_stats()
+    assert "spidermon/validation/items/errors" in stats
```
Comment (ejulio) [inline tests/contrib/scrapy/test_pipeline_cerberus.py]: This test looks like a unwrapped version (specific scenario) of the above.
Is this expected or you wanted to test something else?

**Review by ejulio — CHANGES_REQUESTED** [review]

File: spidermon/contrib/scrapy/pipelines.py (hunk 7 above)
Comment (vipulgupta2048) [reply spidermon/contrib/scrapy/pipelines.py:60]: Actually, the name was originally suggested by you only, I just kept on using it :speak_no_evil:  
But, I like this idea of yours. It's good that we brought it up because I remember there was a project card regarding the same as well. [Here it is.](https://github.com/vipulgupta2048/spidermon/projects/1#card-23531388)

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 8 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: I apologize again regarding this, I have already removed all my comments in the latest commit in my local. I take extra care now to not push any comments that are not needed. Will get rid of it.

File: spidermon/contrib/scrapy/pipelines.py (hunk 7 above)
Comment (ejulio) [reply spidermon/contrib/scrapy/pipelines.py:60]: yup, I recall it was our first idea about this name.
Just thinking there might be better ones.
Or even, a better API for our users :D

File: spidermon/contrib/scrapy/pipelines.py (hunk 1 above)
Comment (vipulgupta2048) [reply spidermon/contrib/scrapy/pipelines.py:104]: @ejulio do review this again. I spent about half the weekend to properly think and implement this. Let me explain what, why and where. 

- The `six.string_types` validation is actually for `URL's` and `paths`, that when `True` data can be sent inside the `get_schema_from()` method. This is placed outside in `pipelines.py` just because I wanted the things going inside the `CerberusValidator()` class to be just what we want and nothing else. This way testing would become much easier for us. As we need to test only schemas and not 15 other things. 

- Now, regarding our discussion of someone picking `CerberusValidator()` and using it somewhere else. Our purpose was to keep it independent. I think this achieves it as well. If you give `Mapping` type data in accordance with the Cerberus schema then this class would work. But, if you like to support URL and paths then the helper function `get_schema_from()` will have to be used. 

- The `Mapping` validation inside the `CerberusValidator()` stays there only. Because we need to check whatever schema that comes is in proper structure or not. However, we would need a lot of testing to see if this works. But, I am feeling good about this. Let me know what you think. I may have forgotten a thing or two since I am coming back to this discussion after a long time. Let me know.

File: spidermon/contrib/monitors/mixins/validation.py [hunk 10]
```diff
@@ -101,8 +101,11 @@ def _get_all_fields(self):
```
Comment (vipulgupta2048) [inline spidermon/contrib/monitors/mixins/validation.py]: Hi reviewers, 
You can skip this file, since its just black formatted.

File: spidermon/contrib/actions/email/__init__.py [hunk 11]
```diff
@@ -22,7 +22,7 @@ class SendEmail(ActionWithTemplates):
     body_text = None
     body_text_template = None
     body_html = None
-    body_html_template = 'reports/email/monitors/result.jinja'
+    body_html_template = "reports/email/monitors/result.jinja"
```
Comment (vipulgupta2048) [inline spidermon/contrib/actions/email/__init__.py]: Hi reviewers,
You can skip this file, since its just black formatted.

File: docs/source/conf.py [hunk 12]
```diff
@@ -18,7 +18,7 @@
 # If extensions (or modules to document with autodoc) are in another directory,
 # add these directories to sys.path here. If the directory is relative to the
 # documentation root, use os.path.abspath to make it absolute, like shown here.
-sys.path.insert(0, os.path.abspath('../..'))
+sys.path.insert(0, os.path.abspath("../.."))
```
Comment (vipulgupta2048) [inline docs/source/conf.py]: Hi reviewers,
You can skip this file, since its just black formatted.

**Review by vipulgupta2048 — COMMENTED** [review]
Hi, here are new replies.

**Review by ejulio — APPROVED** [review]

**Review by vipulgupta2048 — COMMENTED** [review]
Ready to go :rocket: 
Looking forward to any and all final changes suggested @rennerocha @ejulio 
Let's bring this home.

File: tests/contrib/scrapy/test_pipeline_cerberus.py [hunk 13]
```diff
@@ -135,4 +143,58 @@ def test_validation_from_url(mocker):
     pipe.process_item(test_item, None)
 
     assert "spidermon/validation/items/errors" in pipe.stats.stats.get_stats()
-    assert pipe.stats.stats.get_stats().get("spidermon/validation/validators/testitem/cerberus", True)
+    assert pipe.stats.stats.get_stats().get(
+        "spidermon/validation/validators/testitem/cerberus", True
+    )
+
+
+@pytest.mark.skipif(
```
Comment (ejulio) [inline tests/contrib/scrapy/test_pipeline_cerberus.py]: Are you sure about this?

File: tests/contrib/scrapy/test_pipeline_cerberus.py [hunk 14]
```diff
@@ -135,4 +143,58 @@ def test_validation_from_url(mocker):
...
+    sys.version_info < (3, 4), reason="mock requires python3.4 or higher"
+)
+def test_pipelines_with_cerberus_mocked_true_case(mocker):
+    mock_cerberus = mock.patch("spidermon.contrib.scrapy.pipelines.CerberusValidator")
+
+    class Cerberus(object):
```
Comment (ejulio) [inline tests/contrib/scrapy/test_pipeline_cerberus.py]: No need for this

File: tests/contrib/scrapy/test_pipeline_cerberus.py [hunk 15]
```diff
@@ -135,4 +143,58 @@ def test_validation_from_url(mocker):
...
+        def validate(self):
+            return (False, {})
+
+    mock_cerberus.return_value = Cerberus()
```
Comment (ejulio) [inline tests/contrib/scrapy/test_pipeline_cerberus.py]: `mock_cerberus` is already a mock, so
```
mock_cerberus.validate.return_value = (False, {})
```
should be fine.

Besides that, when patching, try to use `autospec`
https://docs.python.org/3/library/unittest.mock.html#patch

File: tests/contrib/scrapy/test_pipeline_cerberus.py [hunk 16]
```diff
@@ -135,4 +143,58 @@ def test_validation_from_url(mocker):
...
+    settings = {SETTING_CERBERUS: [cerberus_test_schema]}
+    test_item = TestItem({"url": "https://github.com"})
+    crawler = get_crawler(settings_dict=settings)
+    pipe = ItemValidationPipeline.from_crawler(crawler)
+    pipe.process_item(test_item, None)
+
+    assert STATS_ITEM_ERRORS not in pipe.stats.stats.get_stats()
+    assert pipe.stats.stats.get_stats().get(
```
Comment (ejulio) [inline tests/contrib/scrapy/test_pipeline_cerberus.py]: This is not good.
Why is the default value `True`?
If the pipeline is required to fill this stats, and it does not, this test will pass, which is a problem

File: tests/contrib/scrapy/test_pipeline_cerberus.py [hunk 17]
```diff
@@ -135,4 +143,58 @@ def test_validation_from_url(mocker):
...
+
+    assert STATS_ITEM_ERRORS in pipe.stats.stats.get_stats()
+    assert pipe.stats.stats.get_stats().get(
+        "spidermon/validation/validators/testitem/cerberus", True
+    )
+    assert (
+        pipe.stats.stats.get_stats()[
+            "spidermon/validation/fields/errors/missing_required_field"
+        ]
+        is 1
```
Comment (ejulio) [inline tests/contrib/scrapy/test_pipeline_cerberus.py]: Test for equality with `==`

**Review by ejulio — CHANGES_REQUESTED** [review]

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 13 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: `Requires: Python >=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*`
Source: [PyPI](https://pypi.org/project/pytest-mock/)

Since Python2 goes out of force soon, hence I think this is a good point that I picked from other tests that I read online and in JSONSchema too.

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 16 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: Oh, I was just testing this bit. I must have forgotten to change this. this actually defaults to false.

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 14 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: Resolving.

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 16 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: Fixed.

File: tests/contrib/scrapy/test_pipeline_cerberus.py [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+    settings = {SETTING_CERBERUS: [cerberus_test_schema]}
+    crawler = get_crawler(settings_dict=settings)
+    pipe = ItemValidationPipeline.from_crawler(crawler)
+    pipe.process_item(item, None)
+    assert pipe.stats.stats.get_stats()["spidermon/validation/validators"] is 1
+    assert (
+        pipe.stats.stats.get_stats()[
+            "spidermon/validation/fields/errors/missing_required_field"
+        ]
+        is 1
```
Comment (ejulio) [inline tests/contrib/scrapy/test_pipeline_cerberus.py]: `==`

Comment (ejulio) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: Also, it looks like `pipe.stats.stats.get_stats()` is used quite a lot.
Maybe a helper method in this test would make it easier to read

**Review by ejulio — CHANGES_REQUESTED** [review]

File: tests/contrib/scrapy/test_pipeline_cerberus.py [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+        ),
+    ],
+)
+def test_stats_not_in_pipeline(item, settings, cases):
+    crawler = get_crawler(settings_dict=settings)
+    pipe = ItemValidationPipeline.from_crawler(crawler)
+    pipe.process_item(item, None)
+    for case in cases:
+        casechecker = lambda x: x not in pipe.stats.stats.get_stats()
+        assert casechecker(case)
```
Comment (ejulio) [inline tests/contrib/scrapy/test_pipeline_cerberus.py]: why not `assert case not in pipe.stats.stats.get_stats()`?

File: tests/contrib/scrapy/test_pipeline_cerberus.py [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+        ]
+        is 1
+    )
+
+
+@pytest.mark.skipif(
+    sys.version_info < (3, 4), reason="mock requires python3.4 or higher"
+)
+def test_validation_from_url(mocker):
+    mocked_get_contents = mocker.patch(
```
Comment (ejulio) [inline tests/contrib/scrapy/test_pipeline_cerberus.py:134]: It looks like the mocking did not work

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 18 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: I apologize @ejulio just missed it. 
Also, I will replace `pipe...get_stats()` with a small function.

Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: I think we would like it a lot now, the helper function now forms a complete sentence.

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 19 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: Dang it, I should make a habit of re-reading my own code, to idiot-proof it.

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 20 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py:134]: It didn't. But, will give it a shot again

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 17 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: Done.

File: tests/contrib/scrapy/test_pipeline_cerberus.py [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+            id="Missing required fields",
+        ),
+    ],
+)
+def test_stats_in_pipeline(item, settings, cases):
+    crawler = get_crawler(settings_dict=settings)
+    pipe = ItemValidationPipeline.from_crawler(crawler)
+    pipe.process_item(item, None)
+    for case in cases:
+        casechecker = lambda x: x in the_stats(pipe)
```
Comment (ejulio) [inline tests/contrib/scrapy/test_pipeline_cerberus.py]: why not `assert x in the_stats(pipe)`

File: tests/contrib/scrapy/test_pipeline_cerberus.py [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+STATS_TYPES = "spidermon/validation/validators/{}/{}"
+SETTING_CERBERUS = "SPIDERMON_VALIDATION_CERBERUS"
+
+
+def assert_in_cerberus_stats(obj):
+    return "{}".format(STATS_TYPES.format(obj.__name__.lower(), "cerberus"))
+
+
+def the_stats(pipe):
+    return pipe.stats.stats.get_stats(pipe)
```
Comment (ejulio) [inline tests/contrib/scrapy/test_pipeline_cerberus.py]: From what I remember, the call in the tests was only `pipe.stats.stats.get_stats()` without `pipe` argument.
If that is accurate, why the change?

**Review by ejulio — CHANGES_REQUESTED** [review]

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 21 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: Right regarding this, in some cases, we are testing 2 even 3 items to be present in stats. That's why I thought Lambda would be a great option. Changed.

File: tests/contrib/scrapy/test_pipeline_cerberus.py (hunk 22 above)
Comment (vipulgupta2048) [reply tests/contrib/scrapy/test_pipeline_cerberus.py]: My bad, I actually was using the `ctrl+d` in VScode to change all duplicate entries to `the_stats(pipe)` which might have to lead it to add one here as well. Fixed. How about now.

**Review by ejulio — APPROVED** [review]

**Review by vipulgupta2048 — COMMENTED** [review]
Well, folks. I guess this is it.
